File: Console/Classes/PreferencesWindowController.m
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Jean-David Gadina - www-xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#import "PreferencesWindowController.h"
#import "Preferences.h"

#define HEXCOLOR( c, a ) [ NSColor colorWithDeviceRed: ( ( CGFloat )( ( c >> 16 ) & 0x0000FF ) ) / ( CGFloat )255  \
                                   green:              ( ( CGFloat )( ( c >>  8 ) & 0x0000FF ) ) / ( CGFloat )255  \
                                   blue:               ( ( CGFloat )( ( c       ) & 0x0000FF ) ) / ( CGFloat )255  \
                                   alpha:              ( CGFloat )a                                                \
                         ]

NS_ASSUME_NONNULL_BEGIN

@interface PreferencesWindowController()

@property( atomic, readwrite, strong ) NSString * fontDescription;
@property( atomic, readwrite, strong ) NSColor  * backgroundColor;
@property( atomic, readwrite, strong ) NSColor  * foregroundColor;
@property( atomic, readwrite, assign ) NSInteger  selectedTheme;

- ( IBAction )chooseFont: ( id )sender;
- ( void )changeFont: ( id )sender;

@end

NS_ASSUME_NONNULL_END

@implementation PreferencesWindowController

- ( instancetype )init
{
    return [ self initWithWindowNibName: NSStringFromClass( self.class ) ];
}

- ( void )windowDidLoad
{
    [ super windowDidLoad ];
    
    [ [ Preferences sharedInstance ] addObserver: self forKeyPath: NSStringFromSelector( @selector( fontName ) ) options: NSKeyValueObservingOptionNew context: NULL ];
    [ [ Preferences sharedInstance ] addObserver: self forKeyPath: NSStringFromSelector( @selector( fontSize ) ) options: NSKeyValueObservingOptionNew context: NULL ];
    
    [ self addObserver: self forKeyPath: NSStringFromSelector( @selector( backgroundColor ) ) options: NSKeyValueObservingOptionNew context: NULL ];
    [ self addObserver: self forKeyPath: NSStringFromSelector( @selector( foregroundColor ) ) options: NSKeyValueObservingOptionNew context: NULL ];
    [ self addObserver: self forKeyPath: NSStringFromSelector( @selector( selectedTheme   ) ) options: NSKeyValueObservingOptionNew context: NULL ];
    
    self.fontDescription = [ NSString stringWithFormat: @"%@ %.0f", [ Preferences sharedInstance ].fontName, [ Preferences sharedInstance ].fontSize ];
    self.backgroundColor = [ NSColor colorWithDeviceRed: [ Preferences sharedInstance ].backgroundColorR green: [ Preferences sharedInstance ].backgroundColorG blue: [ Preferences sharedInstance ].backgroundColorB alpha: 1.0 ];
    self.foregroundColor = [ NSColor colorWithDeviceRed: [ Preferences sharedInstance ].foregroundColorR green: [ Preferences sharedInstance ].foregroundColorG blue: [ Preferences sharedInstance ].foregroundColorB alpha: 1.0 ];
}

- ( void )dealloc
{
    [ [ Preferences sharedInstance ] removeObserver: self forKeyPath: NSStringFromSelector( @selector( fontName ) ) ];
    [ [ Preferences sharedInstance ] removeObserver: self forKeyPath: NSStringFromSelector( @selector( fontSize ) ) ];
    
    [ self removeObserver: self forKeyPath: NSStringFromSelector( @selector( backgroundColor ) ) ];
    [ self removeObserver: self forKeyPath: NSStringFromSelector( @selector( foregroundColor ) ) ];
    [ self removeObserver: self forKeyPath: NSStringFromSelector( @selector( selectedTheme ) ) ];
}

- ( void )observeValueForKeyPath: ( NSString * )keyPath ofObject: ( id )object change: ( NSDictionary * )change context: ( void * )context
{
    if
    (
        object == [ Preferences sharedInstance ]
        &&
        (
               [ keyPath isEqualToString: NSStringFromSelector( @selector( fontName ) ) ]
            || [ keyPath isEqualToString: NSStringFromSelector( @selector( fontSize ) ) ]
        )
    )
    {
        self.fontDescription = [ NSString stringWithFormat: @"%@ %.0f", [ Preferences sharedInstance ].fontName, [ Preferences sharedInstance ].fontSize ];
    }
    else if( object == self && [ keyPath isEqualToString: NSStringFromSelector( @selector( selectedTheme ) ) ] )
    {
        switch( self.selectedTheme )
        {
            case 1:
                
                self.backgroundColor = HEXCOLOR( 0xFFFFFF, 1.0 );
                self.foregroundColor = HEXCOLOR( 0x000000, 1.0 );
                
                break;
                
            case 2:
                
                self.backgroundColor = HEXCOLOR( 0x000000, 1.0 );
                self.foregroundColor = HEXCOLOR( 0xFFFFFF, 1.0 );
                
                break;
                
            case 3:
                
                self.backgroundColor = HEXCOLOR( 0xFFFCE5, 1.0 );
                self.foregroundColor = HEXCOLOR( 0xC3741C, 1.0 );
                
                break;
                
            case 4:
                
                self.backgroundColor = HEXCOLOR( 0x161A1D, 1.0 );
                self.foregroundColor = HEXCOLOR( 0xBFBFBF, 1.0 );
                
                break;
                
            default:
                
                break;
        }
    }
    else if( object == self && [ keyPath isEqualToString: NSStringFromSelector( @selector( backgroundColor ) ) ] )
    {
        {
            CGFloat r;
            CGFloat g;
            CGFloat b;
            
            [ [ self.backgroundColor colorUsingColorSpaceName: NSDeviceRGBColorSpace ] getRed: &r green: &g blue: &b alpha: NULL ];
            
            [ Preferences sharedInstance ].backgroundColorR = r;
            [ Preferences sharedInstance ].backgroundColorG = g;
            [ Preferences sharedInstance ].backgroundColorB = b;
        }
    }
    else if( object == self && [ keyPath isEqualToString: NSStringFromSelector( @selector( foregroundColor ) ) ] )
    {
        {
            CGFloat r;
            CGFloat g;
            CGFloat b;
            
            [ [ self.foregroundColor colorUsingColorSpaceName: NSDeviceRGBColorSpace ] getRed: &r green: &g blue: &b alpha: NULL ];
            
            [ Preferences sharedInstance ].foregroundColorR = r;
            [ Preferences sharedInstance ].foregroundColorG = g;
            [ Preferences sharedInstance ].foregroundColorB = b;
        }
    }
    else
    {
        [ super observeValueForKeyPath: keyPath ofObject: object change: change context: context ];
    }
}

- ( IBAction )chooseFont: ( id )sender
{
    NSFontManager * manager;
    NSFontPanel   * panel;
    NSFont        * font;
    
    font    = [ NSFont fontWithName: [ Preferences sharedInstance ].fontName size: [ Preferences sharedInstance ].fontSize ];
    manager = [ NSFontManager sharedFontManager ];
    panel   = [ manager fontPanel: YES ];
    
    [ manager setSelectedFont: font isMultiple: NO ];
    [ panel makeKeyAndOrderFront: sender ];
}

- ( void )changeFont: ( id )sender
{
    NSFontManager * manager;
    NSFont        * font;
    
    if( [ sender isKindOfClass: [ NSFontManager class ] ] == NO )
    {
        return;
    }
    
    manager = ( NSFontManager * )sender;
    font    = [ manager convertFont: [ manager selectedFont ] ];
    
    [ Preferences sharedInstance ].fontName = font.fontName;
    [ Preferences sharedInstance ].fontSize = font.pointSize;
}

@end

File: Console/Classes/PreferencesWindowController.swift
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2017 Jean-David Gadina - www.xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

import Cocoa

@objc class PreferencesWindowController: NSWindowController
{
    @objc private dynamic var fontDescription: String                    = ""
    @objc private dynamic var backgroundColor: NSColor                   = NSColor.white
    @objc private dynamic var foregroundColor: NSColor                   = NSColor.black
    @objc private dynamic var selectedTheme:   Int                       = 0
    @objc private dynamic var observations:    [ NSKeyValueObservation ] = []
    
    override var windowNibName: NSNib.Name?
    {
        return NSNib.Name( NSStringFromClass( type( of: self ) ) )
    }
    
    override func windowDidLoad()
    {
        super.windowDidLoad()
        
        self.fontDescription = String( format: "%@ %.0f", Preferences.shared.fontName ?? "-", Preferences.shared.fontSize )
        self.backgroundColor = NSColor( deviceRed: Preferences.shared.backgroundColorR, green: Preferences.shared.backgroundColorG, blue: Preferences.shared.backgroundColorB, alpha: 1.0 )
        self.foregroundColor = NSColor( deviceRed: Preferences.shared.foregroundColorR, green: Preferences.shared.foregroundColorG, blue: Preferences.shared.foregroundColorB, alpha: 1.0 )
        
        let o1 = Preferences.shared.observe( \.fontName )
        {
            ( o, c ) in self.fontDescription = String( format: "%@ %.0f", o.fontName ?? "-", o.fontSize )
        }
        
        let o2 = Preferences.shared.observe( \.fontSize )
        {
            ( o, c ) in self.fontDescription = String( format: "%@ %.0f", o.fontName ?? "-", o.fontSize )
        }
        
        let o3 = self.observe( \.backgroundColor )
        {
            ( o, c ) in
            
            var r: CGFloat = 0.0
            var g: CGFloat = 0.0
            var b: CGFloat = 0.0
            
            self.backgroundColor.usingColorSpace( .deviceRGB )?.getRed( &r, green: &g, blue: &b, alpha: nil )
            
            Preferences.shared.backgroundColorR = r
            Preferences.shared.backgroundColorG = g
            Preferences.shared.backgroundColorB = b
        }
        
        let o4 = self.observe( \.foregroundColor )
        {
            ( o, c ) in
            
            var r: CGFloat = 0.0
            var g: CGFloat = 0.0
            var b: CGFloat = 0.0
            
            self.foregroundColor.usingColorSpace( .deviceRGB )?.getRed( &r, green: &g, blue: &b, alpha: nil )
            
            Preferences.shared.foregroundColorR = r
            Preferences.shared.foregroundColorG = g
            Preferences.shared.foregroundColorB = b
        }
        
        let o5 = self.observe( \.selectedTheme )
        {
            ( o, c ) in
            
            switch( self.selectedTheme )
            {
                case 1:
                    
                    self.backgroundColor = self.hexColor( 0xFFFFFF, alpha: 1.0 );
                    self.foregroundColor = self.hexColor( 0x000000, alpha: 1.0 );
                    
                    break;
                    
                case 2:
                    
                    self.backgroundColor = self.hexColor( 0x000000, alpha: 1.0 );
                    self.foregroundColor = self.hexColor( 0xFFFFFF, alpha: 1.0 );
                    
                    break;
                    
                case 3:
                    
                    self.backgroundColor = self.hexColor( 0xFFFCE5, alpha: 1.0 );
                    self.foregroundColor = self.hexColor( 0xC3741C, alpha: 1.0 );
                    
                    break;
                    
                case 4:
                    
                    self.backgroundColor = self.hexColor( 0x161A1D, alpha: 1.0 );
                    self.foregroundColor = self.hexColor( 0xBFBFBF, alpha: 1.0 );
                    
                    break;
                    
                default:
                    
                    break;
            }
        }
        
        self.observations.append( contentsOf: [ o1, o2, o3, o4, o5 ] )
    }
    
    private func hexColor( _ hex: UInt, alpha: CGFloat ) -> NSColor
    {
        let r: CGFloat = CGFloat( ( ( hex >> 16 ) & 0x0000FF ) ) / 255.0
        let g: CGFloat = CGFloat( ( ( hex >>  8 ) & 0x0000FF ) ) / 255.0
        let b: CGFloat = CGFloat( ( ( hex       ) & 0x0000FF ) ) / 255.0
        
        return NSColor( deviceRed: r, green: g, blue: b, alpha: alpha )
    }
    
    @IBAction func chooseFont( _ sender: Any? )
    {
        let font    = NSFont( name: Preferences.shared.fontName as String? ?? "", size: Preferences.shared.fontSize )
        let manager = NSFontManager.shared
        let panel   = manager.fontPanel( true )
        
        if( font != nil )
        {
            manager.setSelectedFont( font!, isMultiple: false )
        }
        
        panel?.makeKeyAndOrderFront( sender )
    }
    
    @IBAction override func changeFont( _ sender: Any? )
    {
        guard let manager = ( sender as AnyObject? ) as? NSFontManager else
        {
            return
        }
        
        guard let selected = manager.selectedFont else
        {
            return
        }
        
        let font = manager.convert( selected )
        
        Preferences.shared.fontName = font.fontName
        Preferences.shared.fontSize = font.pointSize
    }
}

